evm/src/executor/abi.rs

```diff
@@ -680,6 +681,291 @@ pub static HARDHAT_CONSOLE_SELECTOR_PATCHES: Lazy<HashMap<Selector, Selector>> =
     ])
 });
 
+enum FormatSpec {
+    String,  // %s
+    Number,  // %d
+    Integer, // %i
+    Object,  // %o
+}
+
+// FormatValue specifies how a value type is to be formatted
+trait FormatValue: UIfmt {
+    // Formats a value according to the FormatSpec
+    fn fmt(&self, spec: FormatSpec) -> String;
+}
+
+impl FormatValue for String {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.clone(),
+            FormatSpec::Object => format!("'{}'", self.clone()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+impl FormatValue for bool {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number => (*self as i32).to_string(),
+            FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+impl FormatValue for U256 {
+    fn fmt(&self, _spec: FormatSpec) -> String {
+        self.pretty()
+    }
+}
+
+impl FormatValue for I256 {
+    fn fmt(&self, _spec: FormatSpec) -> String {
+        self.pretty()
+    }
+}
+
+impl FormatValue for Address {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+impl FormatValue for Bytes {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+// Formats a `specstr` using the input values, v1, v2 and v3.
```

Comment:
let's turn this into rustdocs with `///` instead of comments `//`

Reply:
yup. Updated the rustdoc.